pkg/controller/scansettingbinding/scansettingbinding_controller.go

```diff
@@ -0,0 +1,515 @@
+var _ reconcile.Reconciler = &ReconcileScanSettingBinding{}
+
+// ReconcileScanSettingBinding reconciles a ScanSettingBinding object
+type ReconcileScanSettingBinding struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// FIXME: generalize for other controllers?
+func (r *ReconcileScanSettingBinding) Eventf(object runtime.Object, eventtype, reason, messageFmt string, args ...interface{}) {
+	if r.recorder == nil {
+		return
+	}
+
+	r.recorder.Eventf(object, eventtype, reason, messageFmt, args...)
+}
+
+func (r *ReconcileScanSettingBinding) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ScanSettingBinding")
+
+	// Fetch the ScanSettingBinding instance
+	instance := &compliancev1alpha1.ScanSettingBinding{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	suite := compliancev1alpha1.ComplianceSuite{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.Name,
+			Namespace: instance.Namespace,
+		},
+		Spec: compliancev1alpha1.ComplianceSuiteSpec{},
+	}
+
+	// Set SettingBinding as the owner of the Suite
+	if err := controllerutil.SetControllerReference(instance, &suite, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var nodeProduct string
+	for _, reference := range instance.Profiles {
+		scan, product, err := newCompScanFromBindingProfile(r, instance, &reference, log)
+		if err != nil {
+			return common.ReturnWithRetriableError(reqLogger, err)
+		}
+
+		if product != "" {
+			if nodeProduct == "" {
+				nodeProduct = product
+			} else {
+				r.Eventf(
+					instance, corev1.EventTypeWarning, "MultipleProducts",
+					"ScanSettingBinding defines multiple products: %s and %s", product, nodeProduct,
+				)
+				// Don't requeue in this case, nothing we can do
+				return reconcile.Result{}, nil
+			}
+		}
+
+		suite.Spec.Scans = append(suite.Spec.Scans, *scan)
+	}
+
+	if instance.Constraint != nil {
+		err := applyConstraint(r, instance, &suite, instance.Constraint, log)
+		if err != nil {
+			return common.ReturnWithRetriableError(reqLogger, err)
+		}
+	}
+
+	err = r.client.Create(context.TODO(), &suite)
+	if err == nil {
+		r.Eventf(
+			instance, corev1.EventTypeWarning, "SuiteCreated",
+			"ComplianceSuite %s/%s created", suite.Namespace, suite.Name,
+		)
+	} else if errors.IsAlreadyExists(err) {
+		err = r.client.Update(context.TODO(), &suite)
```

Comment:
you can't do updates from an object that you instantiated as above (I've run into this several times :confused:  ). Updates have to be for objects fetched from the server. This is because the server will fill out the resource version, uid, etc. These are used to handle collisions in objects (optimistic concurrency control).

the logic would need to be something like this:

```
found := compliancev1alpha1.ComplianceSuite{}
err = r.client.Get(context.TODO(), suiteKey, found)
if err != nil && errors.IsNotFound(err) {
    // No object exists let's try to create it
    err = r.client.Create(context.TODO(), &suite)
    ...
// error handling
} else ... {
}

//  The object was found, so let's figure out if it needs an update
if suiteNeedsUpdate(suite) {
    // Copy the versioning and metadata references that we need
    suiteToBeUpdated := found.DeepCopy()
    // Copy the scans from the suite that you were updating and processing already
    // This way we make sure the update gathers all the info we need.
    suiteToBeUpdated.Spec.Scans := suite.Spec.Scans
    // Copy any other parameters that we need to copy
    // e.g.
    suiteToBeUpdated.Spec.ComplianceSuiteSettings := suite.Spec.ComplianceSuiteSettings
    suiteToBeUpdated.Spec.ComplianceScanSettings := suite.Spec.ComplianceScanSettings
    ....
}
```

Reply:
There is a test for the compliance suite which updates the schedule. You could use that to test this out.

Reply:
Updates now work, tested manually so far. I will add a test that makes sure that a Suite can be updated, but the suite controller so far doesn't update scans, so we need to fix that, too

Reply:
I guess you still need to remove this block, cause the same logic (with the corrections) is now in line 168.